src/dotnet/UI.Blazor/Services/HistoryUI/HistoryUI.cs

```diff
@@ -0,0 +1,228 @@
+using ActualChat.UI.Blazor.Module;
+
+namespace ActualChat.UI.Blazor.Services;
+
+public interface IHistoryUIBackend
+{
+    Task OnPopState(string state);
+}
+
+public class AfterLocationChangedHandledEventsArgs : EventArgs
+{
+    public string Location { get; init; } = "";
+    public bool HasLocationChanged { get; init; }
+    public bool IsBackward { get; init; }
+}
+
+// Keeps track of the initial location.
+// Service should be instantiated on app start.
+public class HistoryUI : IHistoryUIBackend, IDisposable
+{
+    private readonly List<HistoryItem> _history = new ();
+    private readonly Task _whenJsRefInitialized;
+    private PendingHistoryItem? _pendingHistoryItem;
+    private DotNetObjectReference<IHistoryUIBackend>? _blazorRef;
+    private IJSObjectReference? _jsRef;
+    private Task _whenLocationChangedHandled;
+    private State _state;
+    private int _historyIndex;
+
+    private BrowserInfo BrowserInfo { get; }
+    private IJSRuntime JS { get; }
+    private ILogger<HistoryUI> Log { get; }
+    private NavigationManager Nav { get; }
+
+    public string InitialLocation { get; }
+    public bool IsInitialLocation { get; private set; }
```

Comment:
can't it be replaed with _history.Length?

Reply:
See a comment to HistoryUI.OnLocationChanged